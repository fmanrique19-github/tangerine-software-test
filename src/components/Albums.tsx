import React, { useEffect} from 'react';
import { MDBCol, MDBListGroup, MDBListGroupItem, MDBJumbotron, MDBBadge } from "mdbreact";
import IAlbums, { IAlbum } from '../models/Albums';
import Spinner from './Spinner';
import runSaga from '../store/saga/injectSaga';
import { getAlbums, deleteAlbums } from '../sagas/AlbumsSaga';

interface IFAlbums {
    albums: IAlbums;
    md: number;
    classAlbumsMobile: string;
    displayPhotos(id: number, title: string): void;
};

const Albums = (props: IFAlbums): JSX.Element => {
    
    const { albums, md, classAlbumsMobile } = props;

    useEffect(() => {

        runSaga(getAlbums);
        
        return (() => {
            runSaga(deleteAlbums);
        });

    }, []);

    const noAlbums = (): JSX.Element => {
        return (
            <MDBJumbotron>
                <h1 className="text-center">There is no albums available</h1>
            </MDBJumbotron>
        );
    }

    const showPhotos = (e: any, id: number, title: string) => {
        e.preventDefault();
        props.displayPhotos(id, title);
    }

    return (
        <MDBCol md={md as any} className={classAlbumsMobile}>
            {
                !albums ? <Spinner /> : (albums.length === 0) ? noAlbums() :
                <MDBJumbotron>
                    <MDBListGroup>
                        <MDBListGroupItem href="#" className="text-center" active key="0"><h3>Albums</h3></MDBListGroupItem>
                        {
                            albums.map((album: IAlbum) => {
                                return(
                                    <MDBListGroupItem href="#" key={ album.id } onClick={(e: any) => showPhotos(e, album.id, album.title)}>
                                        { album.title }
                                        <MDBBadge color="primary float-right" pill>{ album.id }</MDBBadge>
                                    </MDBListGroupItem>
                                );
                            })
                        }
                    </MDBListGroup>
                </MDBJumbotron>
            }
        </MDBCol>
    );
}

export default Albums;
