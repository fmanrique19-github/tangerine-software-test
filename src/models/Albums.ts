export interface IAlbum {
    userId: number,
    id: number,
    title: string,
}

export default interface IAlbums extends Array<IAlbum>{};
