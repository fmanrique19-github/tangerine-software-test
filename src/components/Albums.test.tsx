import React from 'react';
import renderer from 'react-test-renderer';
import Albums from './Albums';
import mockStore from '../tests/helpers/mockStore';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';

jest.mock('../store/saga/injectSaga');

describe('Albums Component', () => {

    it('Should render a spinner', () => {

        const props = {
            albums: null as any,
        };

        const tree = renderer
            .create(<Provider store={mockStore(props)}><Albums {...props} /></Provider>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Should render there is no albums available', () => {

        const props = {
            albums: [],
        };

        const tree = renderer
            .create(<Provider store={mockStore(props)}><Albums {...props} /></Provider>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Should render 3 albums', () => {

        const props = {
            albums: [
                {
                    "userId": 1,
                    "id": 1,
                    "title": "quidem molestiae enim"
                },
                {
                    "userId": 1,
                    "id": 2,
                    "title": "sunt qui excepturi placeat culpa"
                },
                {
                    "userId": 1,
                    "id": 3,
                    "title": "omnis laborum odio"
                },
            ],
        };;

        const tree = renderer
            .create(<Provider store={mockStore(props)}><Router><Albums {...props} /></Router></Provider>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

});
