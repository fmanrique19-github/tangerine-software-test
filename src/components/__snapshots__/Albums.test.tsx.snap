// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Albums Component Should render 3 albums 1`] = `
<div
  className="col"
  data-test="col"
>
  <div
    className="jumbotron"
    data-test="jumbotron"
  >
    <ul
      className="list-group"
      data-test="list-group"
    >
      <a
        className="list-group-item text-center active list-group-item-undefined"
        data-test="list-group-item"
        href="#"
      >
        <h3>
          Albums
        </h3>
      </a>
      <a
        className="list-group-item list-group-item-undefined"
        data-test="list-group-item"
        href="#"
        onClick={[Function]}
      >
        quidem molestiae enim
        <span
          className="badge primary float-right badge-primary float-right badge-pill"
          data-test="badge"
        >
          1
        </span>
      </a>
      <a
        className="list-group-item list-group-item-undefined"
        data-test="list-group-item"
        href="#"
        onClick={[Function]}
      >
        sunt qui excepturi placeat culpa
        <span
          className="badge primary float-right badge-primary float-right badge-pill"
          data-test="badge"
        >
          2
        </span>
      </a>
      <a
        className="list-group-item list-group-item-undefined"
        data-test="list-group-item"
        href="#"
        onClick={[Function]}
      >
        omnis laborum odio
        <span
          className="badge primary float-right badge-primary float-right badge-pill"
          data-test="badge"
        >
          3
        </span>
      </a>
    </ul>
  </div>
</div>
`;

exports[`Albums Component Should render a spinner 1`] = `
<div
  className="col"
  data-test="col"
>
  <div
    className="d-flex justify-content-center"
  >
    <div
      className="spinner-border"
      role="status"
    >
      <span
        className="sr-only"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;

exports[`Albums Component Should render there is no albums available 1`] = `
<div
  className="col"
  data-test="col"
>
  <div
    className="jumbotron"
    data-test="jumbotron"
  >
    <h1
      className="text-center"
    >
      There is no albums available
    </h1>
  </div>
</div>
`;
